worker/caasapplicationprovisioner/application.go

```diff
@@ -158,37 +161,129 @@ func (a *appWorker) loop() error {
 				return a.dead(app)
 			}
 		case <-appChanges:
-			err = a.updateState(app, false)
+			lastReportedStatus, err = a.updateState(app, false, lastReportedStatus)
 			if err != nil {
 				return errors.Trace(err)
 			}
 		case <-replicaChanges:
-			err = a.updateState(app, false)
+			lastReportedStatus, err = a.updateState(app, false, lastReportedStatus)
 			if err != nil {
 				return errors.Trace(err)
 			}
 		}
 	}
 }
 
-func (a *appWorker) updateState(app caas.Application, force bool) error {
+func (a *appWorker) updateState(app caas.Application, force bool, lastReportedStatus map[string]status.StatusInfo) (map[string]status.StatusInfo, error) {
 	// Fetching the units here is to ensure happens-before consistency
 	// on the deletion of units.
 	observedUnits, err := a.facade.Units(a.name)
 	if err != nil {
-		return errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
 	st, err := app.State()
 	if errors.IsNotFound(err) {
 		// Do nothing
 	} else if err != nil {
-		return errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
 	err = a.facade.GarbageCollect(a.name, observedUnits, st.DesiredReplicas, st.Replicas, force)
 	if err != nil {
-		return errors.Trace(err)
+		return nil, errors.Trace(err)
 	}
-	return nil
+	if force {
+		return nil, nil
+	}
+	// TODO: consolidate GarbageCollect and UpdateApplicationUnits into a single call.
+	units, err := app.Units()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	reportedStatus := make(map[string]status.StatusInfo)
+	args := params.UpdateApplicationUnits{
+		ApplicationTag: names.NewApplicationTag(a.name).String(),
+		Status:         params.EntityStatus{},
+	}
+	for _, u := range units {
+		// For pods managed by the substrate, any marked as dying
+		// are treated as non-existing.
+		if u.Dying {
+			continue
+		}
+		unitStatus := u.Status
+		lastStatus, ok := lastReportedStatus[u.Id]
+		reportedStatus[u.Id] = unitStatus
+		if ok {
+			// If we've seen the same status value previously,
+			// report as unknown as this value is ignored.
```

Comment:
This seems very hacky. Can we add a todo to do it better

Reply:
This is copied, but I'll add the TODO here.